### PR TITLE
Migrate embedding to AndroidX

### PR DESCRIPTION
**Pull request by blasten**
Fixes https://github.com/flutter/flutter/issues/39283

Merge this PR after https://github.com/flutter/flutter/pull/52340 has been merged.

### CODE REVIEW

File: tools/androidx/files.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      "androidx.annotation.VisibleForTesting"
+    ]
+  },
+  {
+    "url": "https://maven.google.com/androidx/core/core/1.2.0/core-1.2.0.aar",
+    "out_file_name": "androidx_core.aar",
+    "maven_dependency": "androidx.core:core:1.2.0",
+    "provides": []
+  }
+]
```
Comment (dnfield) [inline tools/androidx/files.json]: nit: newline at EOF

**Comment by blasten** [comment]
@dnfield This shouldn't introduce any issue since we deprecated non-androidx builds in the previous stable release. I regard to versions, we aren't setting any version constraints at this point since Flutter projects aren't using Gradle 6.0 yet.

> As a nit you can ignore if you want: it might be better to move the download logic to CIPD rather than relying on maven. Maven has been pretty stable for us with this though so ... it's probably fine either way. The advantage of CIPD would be that we could more easily express this in our DEPS file and gclient would know how to track it properly.

Sounds like we should have a bug to track that. However, we still need to generate pom files out of files.json, so we would need to keep some of that logic in the engine repo.

**Comment by blasten** [comment]
fyi @xster

**Comment by xster** [comment]
Should we still send out a "breaking change" announcement? I think it was still previously possible to create a host app in Support and use the engine in Support in 1.12 right? Now using jetifier would be necessary in the host app?

File: shell/platform/android/io/flutter/embedding/engine/plugins/shim/ShimPluginRegistry.java [hunk 2]
```diff
@@ -21,14 +21,18 @@
  * plugin API behind the scenes.
  *
  * <p>The following is an example usage of {@code ShimPluginRegistry} within a {@code
- * FlutterActivity}: {@code // Create the FlutterEngine that will back the Flutter UI. FlutterEngine
- * flutterEngine = new FlutterEngine(context);
+ * FlutterActivity}:
  *
- * <p>// Create a ShimPluginRegistry and wrap the FlutterEngine with the shim. ShimPluginRegistry
- * shimPluginRegistry = new ShimPluginRegistry(flutterEngine, platformViewsController);
+ * <pre>
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/plugins/shim/ShimPluginRegistry.java]: :)

**Comment by xster** [comment]
The actual implementation LGTM. Thanks for finding all the matches.

**Comment by blasten** [comment]
@xster I thought we did that already last quarter. Otherwise, we should send the announcement now that flutter/flutter#52340 is checked in.

**Comment by xster** [comment]
What about existing projects? If I had an add-to-app project created with 1.12 and both are on support and I flutter upgrade after this, I'd have to jetify my host app then right? (I think that's probably ok, but we should have some announcement/wiki)

**Comment by tal-mi** [comment]
> What about existing projects? If I had an add-to-app project created with 1.12 and both are on support and I flutter upgrade after this, I'd have to jetify my host app then right? (I think that's probably ok, but we should have some announcement/wiki)

Thats actually not really okay. Project using add-to-app which are currently unable to update to androidX will be broken. As of 1.17 today they actually are.
Is there a way to keep the embedding with support other than just replacing the pom files with ones generated with 1.12?

**Comment by xster** [comment]
@tal-mi would jetifying your project be possible? 
It's not quite possible to mix and match the versions of the Flutter framework vs the Flutter engine vs the engine dependencies. So the best alternative is to stay on 1.12.

**Comment by tal-mi** [comment]
If I create a library that depends on support libraries and a flutter module then I would have to stay on 1.12. I may have clients which will not want to update their apps to AndroidX so the best way for a "flutter build arr" flow is to provide an (androidX true/false) that will link the embedded libs accordingly.

**Comment by xster** [comment]
Thanks for the context. I would suggest 2 approaches:

1- Create 2 libraries. 1 on support with a 1.12 Flutter module and 1 on AndroidX with a 1.17 module.
2- Create an AndroidX + 1.17 Flutter module library only, but your usage instructions will have to ask users to use jetifier. 

> to provide an (androidX true/false) that will link the embedded libs accordingly.

It's not quite just toggling upstream dependencies. The Flutter engine itself consumes support/AndroidX so it would need different engines altogether. And one instance of the framework isn't tested to work with multiple variants of the engine.

**Comment by tal-mi** [comment]
Thanks for the advice.
Ill look into the best solution. 
We must aspire for AndroidX migration in any case.

**Pull request merged**